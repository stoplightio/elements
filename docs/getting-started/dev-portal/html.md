# Elements Dev Portal in HTML

Setting up Elements Dev Portal is a case of loading up some JavaScript and CSS and embedding a Web Component (custom HTML) into the HTML of your new or existing developer portal. This should work anywhere custom HTML can be embedded and this approach is supported by [pretty much all modern browsers](https://caniuse.com/custom-elementsv1).

For now, the only component in the Dev Portal package is `<elements-stoplight-project/>`, which lets users display Stoplight Projects where they want. 

## Usage

1. Load the Elements Dev Portal JavaScript and default CSS. These can go in the `<head>` tag.
  
```html
<script src="https://unpkg.com/@stoplight/elements-dev-portal/web-components.min.js"></script>
<link rel="stylesheet" href="https://unpkg.com/@stoplight/elements-dev-portal/styles.min.css">  
```

2. Find the "Project ID" from the Project Settings view of your Stoplight Project.

> Project Settings can only be viewed by Project Editors or above. Read more about [project permissions](https://docs.stoplight.io/docs/platform/ZG9jOjg1NjcyNzE-manage-project-access#project-roles).

![The project ID can be found on the Project Settings page in a text box after Display Name and Slug](../../images/projectId.png)

3. Embed the web component inside the HTML `<body>` tag, and enter the project ID.

```html
<elements-stoplight-project
  projectId="cHJqOjYyNTgw"
  router="hash"
></elements-stoplight-project>
```

Alternatively, copy and paste this into an `index.html` and open it in a browser.

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>Elements Dev Portal</title>
  
    <script src="https://unpkg.com/@stoplight/elements-dev-portal/web-components.min.js"></script>
    <link rel="stylesheet" href="https://unpkg.com/@stoplight/elements-dev-portal/styles.min.css">
  </head>
  <body>

    <elements-stoplight-project
      projectId="cHJqOjYyNTgw"
      router="hash"
    ></elements-stoplight-project>

  </body>
</html>
```

5. Finally, serve it with an HTTP server. Grab any HTTP server you like, such as the NPM [http-server](https://www.npmjs.com/package/http-server).

```bash
$ npm install -g http-server
$ http-server

Starting up http-server, serving ./
Available on:
  http://127.0.0.1:8080
  http://192.168.1.28:8080
```

Open up [127.0.0.1:8080](http://127.0.0.1:8080) in the browser and there it should be, Elements in all its glory.

## Configuration

See [Dev Portal Configuration Options](dev-portal-options.md). 

## Examples

<!-- title: Hiding Try It and Mocking -->

```html
<elements-stoplight-project
  projectId="cHJqOjYyNTgw"
  router="hash"
  hideTryIt="true"
  hideMocking="true"
></elements-stoplight-project>
```

## Next steps

You can customize things to your heart's content by wrapping this component in other HTML and changing the CSS, like this [Twitter Bootstrap](https://github.com/stoplightio/elements/blob/main/examples/bootstrap/project.html) example.

Once you're done and your docs are live, send a note to [@stoplightio](https://twitter.com/stoplightio) and Stoplight can help you share them with the world.