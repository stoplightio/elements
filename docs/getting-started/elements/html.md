# Elements in HTML

Setting up Elements is a case of loading up some JavaScript and CSS and embedding a Web Component (custom HTML) into the HTML of your new or existing dev portal. This should work anywhere custom HTML can be embedded and this approach is supported by [pretty much all modern browsers](https://caniuse.com/custom-elementsv1).

For now, the only Web component in the Elements package is `<elements-api/>`, which will display API reference documentation for your OpenAPI. Check out the demo for a live example https://elements-demo.stoplight.io.

## Usage

1. Load the Elements web component JavaScript and default CSS. These can go in the `<head>` tag.
  
```html
<script src="https://unpkg.com/@stoplight/elements/web-components.min.js"></script>
<link rel="stylesheet" href="https://unpkg.com/@stoplight/elements/styles.min.css">  
```

2. Find an OpenAPI document to work with. You could create a document using [Stoplight Studio](https://stoplight.io/studio), [generate it from source code](https://stoplight.io/blog/generate-documentation-from-code/), or browse through some [real-world examples](https://apis.guru/browse-apis/).

3. Embed the web component inside the HTML `<body>` tag.

```html
<elements-api
  apiDescriptionUrl="https://raw.githubusercontent.com/stoplightio/studio-demo/master/reference/todos/todo.v1.yaml"
  router="hash"
/>
```

Alternatively, copy and paste this into an `index.html` and open it in a browser.

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>Elements in HTML</title>
  
    <script src="https://unpkg.com/@stoplight/elements/web-components.min.js"></script>
    <link rel="stylesheet" href="https://unpkg.com/@stoplight/elements/styles.min.css">
  </head>
  <body>

    <elements-api
      apiDescriptionUrl="tictactoe.yaml"
      router="hash"
    />

  </body>
</html>
```

For more configuration options, lets look at the properties available.

## Configuration

These properties can be provided as HTML attributes to the web component.

- `apiDescriptionUrl` - OpenAPI document URL, supporting `http://`, `https://`, and documents containing `$ref` to other HTTP(S) documents.
- `basePath` - Helps when using `router: 'history'` but docs are in a subdirectory like `https://example.com/docs/api`.
- `hideTryIt` - Pass `"true"` to hide the "Try It" panel (the interactive API console).
- `hideInternal` - Pass `"true"` to filter out any content which has been marked as internal with `x-internal`.
- `layout` - There are two layouts for Elements:
  - `sidebar` - (default) Three-column design.
  - `stacked` - Everything in a single column, making integrations with existing websites that have their own sidebar or other columns already.
- `logo` - URL to an image that will show as a small square logo next to the title, above the table of contents.
- `router` -  Determines how navigation should work:
  - `history` - (default) uses the HTML5 history API to keep the UI in sync with the URL.
  - `hash` - uses the hash portion of the URL (i.e. window.location.hash) to keep the UI in sync with the URL.
  - `memory` - keeps the history of your "URL" in memory (does not read or write to the address bar).
  - `static` - renders using the StaticRouter which can help rendering pages on the server.

## Examples

<!-- title: Zoom API with a sidebar and memory router -->

```html
<elements-api
  apiDescriptionUrl="https://raw.githubusercontent.com/stoplightio/Public-APIs/master/reference/zoom/openapi.yaml"
  layout="sidebar"
  router="memory"
/>
```

<!-- title: GitHub API with a Stacked Layout -->

```html
<elements-api
  apiDescriptionUrl="https://api.apis.guru/v2/specs/github.com/1.1.4/openapi.yaml"
  layout="stacked"
/>
```

<!-- title: Hiding "Try It" -->

```html
<elements-api
  apiDescriptionUrl="https://api.apis.guru/v2/specs/github.com/1.1.4/openapi.yaml"
  hideTryIt="true"
/>
```