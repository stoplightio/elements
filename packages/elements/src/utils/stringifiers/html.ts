import { Stringifier } from './types';

export const stringifyHTML: Stringifier<string | HTMLDocument> = (html) => {
  if (typeof html === 'object' && html[Symbol.toStringTag] === 'HTMLDocument') {
    return html.documentElement.outerHTML;
  }

  return String(html);
};
