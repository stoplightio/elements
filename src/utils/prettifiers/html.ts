import { stringifyHTML } from '../stringifiers/html';
import { Prettifier } from './types';

export const prettifyHTML: Prettifier<string | HTMLDocument> = async html => {
  const prettier = await import('prettier/standalone');
  const parserHTML = await import('prettier/parser-html');

  return prettier.format(stringifyHTML(html), {
    parser: 'html',
    plugins: [parserHTML],
    htmlWhitespaceSensitivity: 'ignore',
    tabWidth: 2,
    printWidth: 80,
    endOfLine: 'lf',
  });
};
