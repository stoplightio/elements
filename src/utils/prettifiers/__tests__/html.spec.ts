import * as fs from 'fs';
import { join } from 'path';
import { prettifyHTML } from '../html';

describe('HTML Prettifier', () => {
  test('should prettify content', () => {
    return expect(
      prettifyHTML(fs.readFileSync(join(__dirname, '../__fixtures__/index.html'), 'utf-8')),
    ).resolves.toMatchSnapshot();
  });
});
