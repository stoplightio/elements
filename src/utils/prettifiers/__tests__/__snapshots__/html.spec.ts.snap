// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Prettifier should prettify content 1`] = `
"<html>
  <head>
    <style>
      .foo{display:block}
    </style>
  </head>
  <body>
    <div>a</div>
  </body>
</html>
"
`;
